core/src/main/scala/org/apache/spark/deploy/rest/MesosRestServer.scala

```diff
@@ -0,0 +1,189 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.rest
+
+import java.io.File
+import javax.servlet.http.HttpServletResponse
+
+import org.apache.spark.deploy.DriverDescription
+import org.apache.spark.deploy.ClientArguments._
+import org.apache.spark.deploy.Command
+import org.apache.spark.{SparkConf, SPARK_VERSION => sparkVersion}
+import org.apache.spark.util.Utils
+import org.apache.spark.scheduler.cluster.mesos.ClusterScheduler
+import scala.collection.mutable
+import org.apache.spark.deploy.mesos.MesosDriverDescription
+
+/**
+ * A server that responds to requests submitted by the [[RestClient]].
+ * This is intended to be used in Mesos cluster mode only.
+ *
+ * This server responds with different HTTP codes depending on the situation:
+ *   200 OK - Request was processed successfully
+ *   400 BAD REQUEST - Request was malformed, not successfully validated, or of unexpected type
+ *   468 UNKNOWN PROTOCOL VERSION - Request specified a protocol this server does not understand
+ *   500 INTERNAL SERVER ERROR - Server throws an exception internally while processing the request
+ *
+ * The server always includes a JSON representation of the relevant [[SubmitRestProtocolResponse]]
+ * in the HTTP body. If an error occurs, however, the server will include an [[ErrorResponse]]
+ * instead of the one expected by the client. If the construction of this error response itself
+ * fails, the response will consist of an empty body with a response code that indicates internal
+ * server error.
+ *
+ * @param host the address this server should bind to
+ * @param requestedPort the port this server will attempt to bind to
+ * @param masterConf the conf used by the Master
+ * @param scheduler the scheduler that handles driver requests
+ */
+private [spark] class MesosRestServer(
+    host: String,
+    requestedPort: Int,
+    masterConf: SparkConf,
+    scheduler: ClusterScheduler)
+  extends RestServer(
+    host,
+    requestedPort,
+    masterConf,
+    new MesosSubmitRequestServlet(scheduler, masterConf),
+    new MesosKillRequestServlet(scheduler, masterConf),
+    new MesosStatusRequestServlet(scheduler, masterConf)) {}
+
+class MesosSubmitRequestServlet(
+    scheduler: ClusterScheduler,
+    conf: SparkConf)
+  extends SubmitRequestServlet {
+
+  /**
+   * Build a driver description from the fields specified in the submit request.
+   *
+   * This involves constructing a command that launches a mesos framework for the job.
+   * This does not currently consider fields used by python applications since python
+   * is not supported in mesos cluster mode yet.
+   */
+  private def buildDriverDescription(request: CreateSubmissionRequest): MesosDriverDescription = {
```

Comment:
by this way there is actually nothing that restricts Mesos to follow standalone mode's way submitting a `DriverDescription`. In our case it might make more sense to merge `MesosDriverDescription` with `DriverSubmission` since the latter contains the former anyway. Then you can just express `schedulerProperties` as another field in `DriverSubmission` instead of worrying about how to extend `DriverDescription` in a clean way.